packages/back-end/src/util/integrations.ts

```diff
@@ -0,0 +1,73 @@
+import { InformationSchema, RawInformationSchema } from "../types/Integration";
+
+export function formatInformationSchema(
```

Comment:
This method seems to be repeating a lot of the same operations, which may not be a problem if results are small, but we are repeating `.some()` and `.findIndex()` operations where it looks like we don't need to be. You may be able to optimize it with a hashmap/lookup approach rather than iterating through the lists.

This looks like something that would be good to add some tests for—it doesn't make DB calls and returns output based on the provided input. When we need to do things like handle more data source types, this would be handy to ensure we don't add regressions.

Reply:
Great suggestion. I've refactored this to the following.

```ts
export function formatInformationSchema(
  results: RawInformationSchema[],
  datasourceType: DataSourceType
): InformationSchema[] {
  const formattedResultsMap = new Map();

  results.forEach((row) => {
    if (!formattedResultsMap.has(row.table_catalog)) {
      formattedResultsMap.set(row.table_catalog, {
        database_name: row.table_catalog,
        schemas: new Map(),
        path: getPath(datasourceType, {
          table_catalog: row.table_catalog,
        }),
      });
    }

    const currentSchemaCatalog = formattedResultsMap.get(row.table_catalog);

    if (!currentSchemaCatalog.schemas.has(row.table_schema)) {
      currentSchemaCatalog.schemas.set(row.table_schema, {
        schema_name: row.table_schema,
        tables: new Map(),
        path: getPath(datasourceType, {
          table_catalog: row.table_catalog,
          table_schema: row.table_schema,
        }),
      });
    }

    const currentTableSchema = currentSchemaCatalog.schemas.get(
      row.table_schema
    );

    if (!currentTableSchema.tables.has(row.table_name)) {
      currentTableSchema.tables.set(row.table_name, {
        table_name: row.table_name,
        columns: new Map(),
        path: getPath(datasourceType, {
          table_catalog: row.table_catalog,
          table_schema: row.table_schema,
          table_name: row.table_name,
        }),
      });
    }

    const currentColumnsSchema = currentTableSchema.tables.get(row.table_name);

    if (!currentColumnsSchema.columns.has(row.column_name)) {
      currentColumnsSchema.columns.set(row.column_name, {
        column_name: row.column_name,
        data_type: row.data_type,
        path: getPath(datasourceType, {
          table_catalog: row.table_catalog,
          table_schema: row.table_schema,
          table_name: row.table_name,
          column_name: row.column_name,
        }),
      });
    }
  });

  const formattedResultsArr = Array.from(formattedResultsMap.values());

  // Convert everything from maps to arrays.
  formattedResultsArr.forEach((schema) => {
    schema.schemas = Array.from(schema.schemas.values());
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    schema.schemas.forEach((table: any) => {
      table.tables = Array.from(table.tables.values());
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      table.tables.forEach((column: any) => {
        column.columns = Array.from(column.columns.values());
      });
    });
  });

  return formattedResultsArr;
}
```

I'm currently converting everything back to an array before returning - this feels like an extra step, but I haven't seen us store map objects in Mongo yet, so I did this to keep the pattern, but open to feedback.